include/aspect/particle/particle.h

```diff
@@ -21,119 +21,141 @@
+         * origin.
          */
-        BaseParticle ();
+        Particle ();
 
         /**
-         * Constructor for BaseParticle, creates a particle with the specified
-         * ID at the specified location with zero velocity. Note that Aspect
-         * does not check for duplicate particle IDs so the user must be sure
-         * the IDs are unique over all processes.
+         * Constructor for Particle, creates a particle with the specified
+         * ID at the specified location. Note that Aspect
+         * does not check for duplicate particle IDs so the generator must
+         * make sure the IDs are unique over all processes.
          *
          * @param[in] new_loc Initial location of particle.
          * @param[in] new_id Globally unique ID number of particle.
          */
-        BaseParticle (const Point<dim> &new_loc,
-                      const double &new_id);
+        Particle (const Point<dim> &new_loc,
+                  const particle_index &new_id);
+
+        /**
+         * Constructor for Particle, creates a particle from a data vector.
+         * This constructor is usually called after sending a particle to a
+         * different process.
+         *
+         * @param[in,out] begin_data First data component.
+         * @param[in] data_len Number of components of the begin_data vector
+         * that will be read in by this particle.
+         */
+        Particle (const void *&begin_data,
+                  const unsigned int data_len);
 
         /**
-         * Destructor for BaseParticle
+         * Destructor for Particle
          */
         virtual
-        ~BaseParticle ();
+        ~Particle ();
 
         /**
          * Get the number of doubles required to represent this particle for
-         * communication.
+         * communication. This includes the base properties like position
+         * and id as well as the additional user requested properties.
          *
          * @return Number of doubles required to represent this particle
          */
-        static unsigned int
-        data_len ();
+        unsigned int
+        data_len () const;
 
         /**
-         * Read the particle data from the specified vector of doubles.
-         *
-         * @param [in] data The vector of double data to read from.
-         * @param [in] pos The position in the data vector to start reading
-         * from.
-         * @return The position in the vector of the next unread double.
-         */
-        virtual unsigned int read_data(const std::vector<double> &data, const unsigned int &pos);
+          * Set the number of doubles required to represent this particle for
+          * communication. This includes the base properties position
+          * and id as well as the additional user requested properties.
+          *
+          * @param [in] data_len Number of doubles to represent this particle
+          */
+        void
+        set_data_len (const unsigned int data_len);
```

Comment:
I have changed the name to `set_n_property_components` to make it more descriptive, and it is called from the property manager class to ensure that the vector containing this particle's properties is of the correct size to hold its properties. I have added this comment to the documentation.